File: EmojibleIOS/EmojibleIOS/DataSources/StudentsInClassDataSource.swift
//
//  StudentsInClassDataSource.swift
//  EmojibleIOS
//
//  Created by Furkan Yakal on 19.12.2020.
//

import Foundation
import Firebase

class StudentsInClassDataSource{
    let database = Firestore.firestore()
    var snapshotListener:ListenerRegistration?
    let notificationCenter = NotificationCenter.default
    var students: [StudentModel] = []
    
    func startObservingStudentsInAClass(classId: String){
        let classesCollectionRef = database.collection("Users").whereField("classId", isEqualTo: classId)
        snapshotListener = classesCollectionRef.addSnapshotListener{ [unowned self] querySnapshot, error in
            guard let documents = querySnapshot?.documents else { return }
            students = documents.map{StudentModel(dictionary: $0.data())}
            notificationCenter.post(name: .studentsInClassChanged, object: nil, userInfo:["studentsInClassChanged":students])
        }
    }
    
    func stopObservingStudentsInAClass(){
        guard let snapshotListener = snapshotListener else { return }
        snapshotListener.remove()
    }
    
    private init(){}
    private static let instance = StudentsInClassDataSource()
    public static func getInstance() -> StudentsInClassDataSource{
        return .instance
    }
}

File: EmojibleAndroid/app/src/main/java/com/dji/emojibleandroid/dataSources/StudentsInClassDataSource.kt
package com.dji.emojibleandroid.dataSources

import com.dji.emojibleandroid.models.StudentModel
import com.dji.emojibleandroid.services.Changes
import com.dji.emojibleandroid.services.NotificationCenter
import com.google.firebase.firestore.ListenerRegistration
import com.google.firebase.firestore.ktx.firestore
import com.google.firebase.ktx.Firebase

class StudentsInClassDataSource private constructor(){
    val database = Firebase.firestore
    var snapshotListener: ListenerRegistration? = null
    val notificationCenter = NotificationCenter.instance
    var students: List<StudentModel>? = null

    fun startObservingStudentsInAClass(classId: String) {
        val classesCollectionRef = database.collection("Users").whereEqualTo("classId", classId)
        snapshotListener = classesCollectionRef.addSnapshotListener { querySnapshot, error ->
            val documents = querySnapshot?.documents ?: return@addSnapshotListener
            students = documents.map { StudentModel(it.data as HashMap<String, Any>) }
            notificationCenter.post(Changes.studentsInClassChanged, null, hashMapOf("studentsInClassChanged" to students))
        }
    }

    fun stopObservingStudentsInAClass() {
        val snapshotListener = snapshotListener ?: return
        snapshotListener.remove()
    }

    private object HOLDER {
        val INSTANCE = StudentsInClassDataSource()
    }

    companion object {
        val instance: StudentsInClassDataSource by lazy { HOLDER.INSTANCE }
    }
}